File: Sources/VihChatBotSDK/Repositories/ChatRepository.swift
import Foundation

/// Mirrors `data/repository/ChatRepository.kt`.
public final class ChatRepository: BaseRepository {

    private let apiService: ApiService

    public init(apiService: ApiService, loaderHost: LoaderHost?) {
        self.apiService = apiService
        super.init(loaderHost: loaderHost)
    }

    public func getChatResponse(
        showBlockingLoader: Bool,
        question: String,
        sessionId: String,
        hashcode: String,
        enterpriseId: String
    ) async throws -> ChatMessageModel {
        try await doSafeAPIRequest(showBlockingLoader: showBlockingLoader) {
            try await self.apiService.getChatResponse(
                hashcode: hashcode,
                enterpriseId: enterpriseId,
                question: question,
                sessionId: sessionId
            )
        }
    }

    public func getChatHistory(
        showBlockingLoader: Bool, channelId: String, enterpriseId: String
    ) async throws -> ChatHistoryModel {
        try await doSafeAPIRequest(showBlockingLoader: showBlockingLoader) {
            try await self.apiService.getChatHistory(channelId: channelId, enterpriseId: enterpriseId)
        }
    }

    public func getChatListResponse(
        showBlockingLoader: Bool, hashcode: String
    ) async throws -> ChatListModelResponse {
        try await doSafeAPIRequest(showBlockingLoader: showBlockingLoader) {
            try await self.apiService.getChatListResponse(hashcode: hashcode)
        }
    }

    public func getEnterpriseDetails(
        showBlockingLoader: Bool, enterpriseId: String
    ) async throws -> EnterpriseApiResponse {
        try await doSafeAPIRequest(showBlockingLoader: showBlockingLoader) {
            try await self.apiService.getEnterprises(enterpriseId: enterpriseId)
        }
    }
}

File: vihChatBot/src/main/java/com/vihmessenger/vihchatbot/data/repository/ChatRepository.kt
package com.vihmessenger.vihchatbot.data.repository

import BaseActivity
import com.vihmessenger.vihchatbot.data.model.ChatHistoryModel
import com.vihmessenger.vihchatbot.data.model.ChatListModelResponse
import com.vihmessenger.vihchatbot.data.model.ChatMessageModel
import com.vihmessenger.vihchatbot.data.model.EnterPriseModel
import com.vihmessenger.vihchatbot.data.model.EnterpriseApiResponse
import com.vihmessenger.vihchatbot.data.services.ApiService

class ChatRepository(
    private val apiService: ApiService, private val baseActivity: BaseActivity?
) : BaseRepository(baseActivity, apiService) {

    suspend fun getChatResponse(
        showBlockingLoader: Boolean,
        question: String,
        sessionId: String,
        hashcode: String,
        enterpriseId: String
    ): ChatMessageModel {
        return doSafeAPIRequest(
            call = { apiService.getChatResponse(hashcode, enterpriseId, question, sessionId) },
            showBlockingLoader = showBlockingLoader
        ) ?: throw NoConnectionException("Failed to get chat response")
    }

    suspend fun getChatHistory(
        showBlockingLoader: Boolean, channelId: String, enterpriseId: String
    ): ChatHistoryModel {
        return doSafeAPIRequest(
            call = {
                apiService.getChatHistory(
                    channelId, enterpriseId
                )
            }, showBlockingLoader = showBlockingLoader
        ) ?: throw NoConnectionException("Failed to get chat history")
    }

    suspend fun getChatListResponse(
        showBlockingLoader: Boolean,
        hashcode: String,
    ): ChatListModelResponse {
        return doSafeAPIRequest(
            call = { apiService.getChatListResponse(hashcode) },
            showBlockingLoader = showBlockingLoader
        ) ?: throw NoConnectionException("Failed to get chat list")
    }

    suspend fun getEnterpriseDetails(
        showBlockingLoader: Boolean,
        enterprise_id: String,
    ): EnterpriseApiResponse {
        return doSafeAPIRequest(
            call = { apiService.getEnterprises(enterprise_id) },
            showBlockingLoader = showBlockingLoader
        ) ?: throw NoConnectionException("Failed to get chat list")
    }
}

class NoConnectionException(message: String) : Exception(message)